components/merkledb/src/access/mod.rs

```diff
@@ -0,0 +1,242 @@
+
+    fn get_or_create_view(
+        &self,
+        addr: IndexAddress,
+        index_type: IndexType,
+    ) -> Result<ViewWithMetadata<Self::Base>, AccessError> {
+        ViewWithMetadata::get_or_create(self.clone(), &addr, index_type).map_err(|e| AccessError {
+            addr,
+            kind: AccessErrorKind::WrongIndexType {
+                expected: index_type,
+                actual: e.index_type(),
+            },
+        })
+    }
+}
+
+/// Access that prepends the specified prefix to each created view.
+///
+/// # Examples
+///
+/// ```
+/// use exonum_merkledb::{access::{AccessExt, Prefixed}, Database, TemporaryDB};
+///
+/// let db = TemporaryDB::new();
+/// let fork = db.fork();
+/// let prefixed = Prefixed::new("prefixed", &fork);
+/// prefixed.get_list("list").extend(vec![1_u32, 2, 3]);
+/// let same_list = fork.as_ref().get_list::<_, u32>("prefixed.list");
+/// assert_eq!(same_list.len(), 3);
+/// ```
+#[derive(Debug, Clone)]
+pub struct Prefixed<'a, T> {
+    access: T,
+    prefix: Cow<'a, str>,
+}
+
+impl<'a, T: Access> Prefixed<'a, T> {
+    /// Creates new prefixed access.
+    ///
+    /// # Panics
+    ///
+    /// Will panic if the prefix does not conform to valid names for indexes.
+    pub fn new(prefix: impl Into<Cow<'a, str>>, access: T) -> Self {
+        let prefix = prefix.into();
+        assert_valid_name(prefix.as_ref());
+        Self { access, prefix }
+    }
+}
+
+impl<T: Access> Access for Prefixed<'_, T> {
+    type Base = T::Base;
+
+    fn get_or_create_view(
+        &self,
+        addr: IndexAddress,
+        index_type: IndexType,
+    ) -> Result<ViewWithMetadata<Self::Base>, AccessError> {
+        let prefixed_addr = addr.prepend_name(self.prefix.as_ref());
+        self.access.get_or_create_view(prefixed_addr, index_type)
+    }
+}
+
+/// Error together with location information.
+#[derive(Debug, Fail)]
+pub struct AccessError {
+    /// Address of the index where the error has occurred.
+    pub addr: IndexAddress,
+    /// Error kind.
+    #[fail(cause)]
+    pub kind: AccessErrorKind,
+}
+
+impl fmt::Display for AccessError {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // TODO: implement `Display` for `IndexAddress` for human-readable errors
+        write!(formatter, "Error accessing {:?}: {}", self.addr, self.kind)
+    }
+}
+
+/// Error that can be emitted during accessing an object from the database.
+#[derive(Debug, Fail)]
+pub enum AccessErrorKind {
+    /// Index has wrong type.
+    #[fail(
+        display = "Wrong index type: expected {:?}, but got {:?}",
+        expected, actual
+    )]
+    WrongIndexType {
+        /// Expected index type.
+        expected: IndexType,
+        /// Actual index type.
+        actual: IndexType,
+    },
+
+    /// Custom error.
+    #[fail(display = "{}", _0)]
+    Custom(#[fail(cause)] Error),
+}
+
+/// Restores an object from the database.
+pub trait Restore<T: Access>: Sized {
```

Comment:
In my opinion, such name can lead to confusion. What actually "restore" means: restore broken index, restore access to the index or something else?

Reply:
Would something like `FromAccess` be acceptable?